core/primitives/src/block_header.rs

```diff
@@ -845,6 +972,41 @@ impl BlockHeader {
         }
     }
 
+    /// Helper function for `new_impl` and `old_impl` to compute the hash and signature of the hash from the block header parts.
+    /// Exactly one of the `signer` and `signature` must be provided.
+    /// If `signer` is given signs the header with given `prev_hash`, `inner_lite`, and `inner_rest` and returns the hash and signature of the header.
+    /// If `signature` is given, uses the signature as is and only computes the hash.  
+    fn compute_hash_and_sign<T>(
+        signer: Option<&ValidatorSigner>,
+        signature: Option<Signature>,
+        prev_hash: CryptoHash,
+        inner_lite: &BlockHeaderInnerLite,
+        inner_rest: &T,
+    ) -> (CryptoHash, Signature)
+    where
+        T: BorshSerialize + ?Sized,
+    {
+        assert_ne!(
+            signer.is_some(),
+            signature.is_some(),
+            "Exactly one of signer and signature must be provided"
+        );
```

Comment:
We should replace them with a single enum then, with two variants, e.g.
```
enum SignatureSource<'a> {
    Signer(&'a ValidatorSigner)
    Signature(Signature)
}
```
which documents itself.

Reply:
done, thanks for the suggestion.